src/main/scala/com/microsoft/hyperspace/index/IndexConfig.scala

```diff
@@ -153,14 +186,51 @@ object IndexConfig {
      * @return an [[IndexConfig]] object.
      */
     def create(): IndexConfig = {
-      IndexConfig(indexName, indexedColumns, includedColumns)
+      IndexConfig(
+        indexName,
+        indexedColumns,
+        IncludedColumnsConfig(columnsToInclude, columnsToExclude))
     }
   }
 
   /**
-   *  Creates new [[IndexConfig.Builder]] for constructing an [[IndexConfig]].
+   * Creates new [[IndexConfig.Builder]] for constructing an [[IndexConfig]].
    *
    * @return an [[IndexConfig.Builder]] object.
    */
   def builder(): Builder = new Builder
 }
+
+/**
+ * Specifies list of included columns in terms of columns to include and/or exclude.
+ *
+ * @param include List of column names to include as include columns.
+ * @param exclude List of column names to exclude to form list of included columns.
+ */
+case class IncludedColumnsConfig(include: Seq[String] = Nil, exclude: Seq[String] = Nil) {
```

Comment:
IMO, having `IncludedColumnsConfig` inside `IndexConfig` is a bit confusing. Any reason to introduce this class? @pirz 

Since this is a part of user APIs, it should look clear and concise.
This is the current API using IncludedColumnsConfig:

```
val indexConfig = IndexConfig("index1", Seq("RGUID"), IncludedColumnsConfig(Seq(), Seq("Date")))
```

I think it's better to accept w/o the additional class, e.g. `IndexConfig(name, indexedColumns, includedColumns, excludedColumns)` and we could create it internally if needed.

@imback82 Could you give any comment on this?
I suggested including all 3 types of columns (indexed/included/excluded) https://github.com/microsoft/hyperspace/pull/309#discussion_r553808258, but it might be better not to use another class in user APIs.

Reply:
@sezruby Lets be clear here: `IncludedColumnsConfig` is not *inside* `IndexConfig`, it is a completely separate case class, both in definition and usage. I added its definition in the same file just to keep it close to `IndexConfig` as they are used together. The definition can be moved to its own separate file.
 
I am not sure what exactly is wrong with adding a new class, `include` and `exclude` lists are both related to the definition of included columns and packing them together makes sense. The API call is clean and clear from a user perspective. I am not sure how unwrapping the config and bringing out two lists and add them separately removes any confusion. Can you be more precise about the *confusion* you are concerned about?

Reply:
In terms of :
* as we always take either of "included" or "excluded", it's confusing to handle them using one class
* they can use both included and excluded but excluded won't be used 
* we don't need to keep "excluded" columns after creation

minor points:
* need to import `IncludedColumnConfig` other than `IndexConfig` in their applications 
* public `lazy val`s will be exposed unnecessarily

We could keep the class internally if needed but I don't think it's good to expose the class in API.

Reply:
Yea, I think we can update `IndexConfig` to have `name`, `indexedColumns`, `includedColumns` and `excludedColumns` similar to the `Builder`? (In this case, it's easier to add a verification check - inside `IndexConfig` - whether any of excluded columns are present in indexed/included columns, as well.)

Later when we need to add more info (if any) to the included columns,  we can come back to refactor? What do you guys think?

Reply:
The reason I have boxed `include` and `exclude` together is because they are related to a single concept in our covering index definition: "Included Columns". 
Removing `IncludedColumnsConfig` class and adding two separate lists `includedColumns` and `excludedColumns` does not add any more clarity that both `include` and `exclude` can not be used together. Whatever check/requirement we had inside `IncludedColumnsConfig` now comes one level up, so I am not sure why you say that clarifies/removes such requirements.

I am not religious about keeping `IncludedColumnsConfig` class, however one side reason that I considered when adding it was avoiding any confusion in naming arguments: In our Covering Index definition, "Included Columns" is a well-defined concept: "The payload columns that piggyback on index records which are physically organized according to indexed columns". Now if we change the API as:
`IndexConfig(name, indexedColumns, includedColumns, excludedColumns)`

Then `includedColumns` and `excludedColumns` arg names seems confusing as the first arg has the same name as the "included columns" concept. However, boxing them together and under the name `IncludedColumnsConfig` clarifies that these two args are used to define "included columns" list for an index. @imback82 If you are absolutely in favor of dropping `IncludedColumnsConfig`, then what do you suggest for arg names?

Reply:
I don't have a strong opinion here. But we need to document either way since we should not expect the users have a prior knowledge on the" included columns" concept, not to mention the covering index.

Also, even if we introduce `IncludedColumnsConfg`, I suggest exposing `apply` that takes these four arguments for the ease of use, as @sezruby pointed out above.

Reply:
@imback82 WDYT about the alternative approach from last comment? (I copied it below).

One more idea (coming from [discussion](https://github.com/microsoft/hyperspace/pull/309#discussion_r560600669)) could be:
As both `include` and `exclude` should not be set together, one option could be using:
1. One Seq[String] () to list column(s) names, and
2. Using a mode to specify include or exclude.

Then we can list the columns that we want to include or exclude (depending on whichever is shorter to add) w.r.t to the DF schema passed to `createIndex` and create our desired "included columns" list. Moreover, with `("exclude", Seq())` we can include all columns and with `("include", Seq())` we can exclude all columns.

Reply:
@pirz +1 for keeping included/excluded columns in a single class, because what really matters at the end of the day is which columns are included for the covering index. Included/excluded column spec from user is merely a mean to determine those final included columns.

Concerning the possibility of invalid spec (both included/excluded columns are set), how about making separate classes for each case?

```scala
sealed trait IncludedColumnsConfig
case class IncludeColumns(columns: Seq[String]) extends IncludedColumnsConfig
case class ExcludeColumns(columns: Seq[String]) extends IncludedColumnsConfig
```

Note that the imperative form is used in the names because they are sort of telling Hyperspace that it should "include" the specified columns only, or "exclude" the specified columns and must include all other columns.